src/Components/Blazor/WebAssembly.Authentication/src/Services/RemoteAuthenticationService.cs

```diff
@@ -0,0 +1,188 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Security.Claims;
+using System.Text.Json;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Components.Authorization;
+using Microsoft.Extensions.Options;
+using Microsoft.JSInterop;
+
+namespace Microsoft.AspNetCore.Components.WebAssembly.Authentication
+{
+    /// <summary>
+    /// The default implementation for <see cref="IRemoteAuthenticationService{TRemoteAuthenticationState}"/> that uses JS interop to authenticate the user.
+    /// </summary>
+    /// <typeparam name="TRemoteAuthenticationState">The state to preserve across authentication operations.</typeparam>
+    /// <typeparam name="TProviderOptions">The options to be passed down to the underlying JavaScript library handling the authentication operations.</typeparam>
+    public class RemoteAuthenticationService<TRemoteAuthenticationState, TProviderOptions> : AuthenticationStateProvider, IRemoteAuthenticationService<TRemoteAuthenticationState>
+         where TRemoteAuthenticationState : RemoteAuthenticationState
+         where TProviderOptions : new()
+    {
+        private bool _initialized = false;
+
+        /// <summary>
+        /// The <see cref="IJSRuntime"/> to use for performing JavaScript interop operations.
+        /// </summary>
+        protected readonly IJSRuntime _jsRuntime;
+
+        /// <summary>
+        /// The options for the underlying JavaScript library handling the authentication operations.
+        /// </summary>
+        protected readonly RemoteAuthenticationOptions<TProviderOptions> _options;
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="RemoteAuthenticationService{TRemoteAuthenticationState, TProviderOptions}"/>.
+        /// </summary>
+        /// <param name="jsRuntime">The <see cref="IJSRuntime"/> to use for performing JavaScript interop operations.</param>
+        /// <param name="options">The options to be passed down to the underlying JavaScript library handling the authentication operations.</param>
+        public RemoteAuthenticationService(
+            IJSRuntime jsRuntime,
+            IOptions<RemoteAuthenticationOptions<TProviderOptions>> options)
+        {
+            _jsRuntime = jsRuntime;
+            _options = options.Value;
+        }
+
+        /// <inheritdoc />
+        public override async Task<AuthenticationState> GetAuthenticationStateAsync() => new AuthenticationState(await GetCurrentUser());
```

Comment:
Other parts of the framework tend to call `GetAuthenticationStateAsync` quite freely and indiscriminately, based on the expectation that normally it will just be returning cached data and not performing any expensive operations.

Would it be practical to cache the current authentication state inside this class, and only invalidate it when some operation occurs that makes us believe the previous info is no longer correct?

Or do we really make a call to some external service to revalidate the user's tokens every time this is called?

Reply:
done